File: 9. Boundary Elements.cpp
// Topic: Boundary Elements
#include <bits/stdc++.h>
using namespace std;
int main()
{
    vector<vector<int>> A = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int n = A.size(), m = A[0].size();
    for (int j = 0; j < m; ++j) cout << A[0][j] << " ";
    for (int i = 1; i < n - 1; ++i) cout << A[i][m - 1] << " ";
    if (n > 1)
        for (int j = m - 1; j >= 0; --j) cout << A[n - 1][j] << " ";
    for (int i = n - 2; i >= 1; --i) cout << A[i][0] << " ";
    cout << "\n";
}

// TC: O(n*m)
// SC: O(1)

File: 9. Boundary Elements.py
"""Topic: Boundary Elements"""
def boundary(A):
    n=len(A); m=len(A[0])
    out=[]
    out.extend(A[0])
    for i in range(1,n-1): out.append(A[i][m-1])
    if n>1: out.extend(reversed(A[n-1]))
    for i in range(n-2,0,-1): out.append(A[i][0])
    return out

if __name__=='__main__':
    A=[[1,2,3],[4,5,6],[7,8,9]]
    print(*boundary(A))

# TC: O(n*m)
# SC: O(n+m)
